lib/tpm2_hierarchy.c

```diff
@@ -0,0 +1,78 @@
+//**********************************************************************;
+// Copyright (c) 2018, Intel Corporation
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of Intel Corporation nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdbool.h>
+
+#include <sapi/tpm20.h>
+
+#include "log.h"
+#include "tpm2_util.h"
+
+bool tpm2_hierarchy_from_optarg(const char *value,
+        TPMI_RH_PROVISION *hierarchy) {
+
+    if (!value) {
+        return false;
+    }
+
+    bool is_o = !strcmp(value, "o");
+    if (is_o) {
+        *hierarchy = TPM2_RH_OWNER;
+        return true;
+    }
+
+    bool is_p = !strcmp(value, "p");
+    if (is_p) {
+        *hierarchy = TPM2_RH_PLATFORM;
+        return true;
+    }
+
+    bool is_e = !strcmp(value, "e");
+    if (is_e) {
+        *hierarchy = TPM2_RH_ENDORSEMENT;
+        return true;
+    }
+
+    bool is_n = !strcmp(value, "n");
+    if (is_n) {
+        *hierarchy = TPM2_RH_NULL;
+        return true;
+    }
+
+    bool result = tpm2_util_string_to_uint32(value, hierarchy);
+    if (!result) {
+        LOG_ERR("Incorrect hierarchy value, got: \"%s\", expected [o|p|e|n]"
+                "or a number",
```

Comment:
I would say: or an hexadecimal number.

Reply:
BTW, should we check that the hex value is a correct TPM_RH_* an report that or let the command fail when is sent by the TPM?

Reply:
My latest comment made me realize that probably we should check for the Handle value passed to the tools too. I've filled issue #880 for that.

Reply:
I'd say let the TPM check that.

Reply:
Well checking is trivial, and I think it would be a better user experience if the tool can catch those and report a user friendly error message instead of getting a more generic error from the TPM.

I'm very busy with some internal projects lately but I'll try to make some time to propose a patch.

Reply:
The more I think about this, if you're putting in the raw hex, you should know what your doing and let it hit the TPM or another layer. If someone is trying to do something odd, perhaps with an experimental TPM feature and needs to pass a value straight in, they should be able to.

I'd like to be mechanism here, not policy.

Reply:
Yes, I agree with you when talking about the hierarchies (we do have a symbolic representation so why anyone would use a hex value?) but I still think it should be nice to have for handles since users have no other option that use hex values.